### PR TITLE
High: fix infloop when "validate-with" schema does not validate

### PR DESCRIPTION
**Pull request by jnpkrn**
Test case:
Change constraints-1.0.rng so as to deny score-attribute
in rsc_colocation; then:

$ make  # so as to generate the schemas
$ PCMK_schema_directory=$(pwd)/xml tools/crm_simulate -x $(pwd)/pengine/test10/use-after-free-merge.xml -S

BEFORE:
Entity: line 56: element rsc_colocation: Relax-NG validity error : Invalid attribute score-attribute for element rsc_colocation
Entity: line 49: element rsc_location: Relax-NG validity error : Element constraints has extra content: rsc_location
[...infloop...]

AFTER:
Entity: line 56: element rsc_colocation: Relax-NG validity error : Invalid attribute score-attribute for element rsc_colocation
Entity: line 49: element rsc_location: Relax-NG validity error : Element constraints has extra content: rsc_location
[several times]
4 of 5 resources DISABLED and 0 BLOCKED from being started due to failures
[...]

### CODE REVIEW

**Comment by kgaillot** [comment]
I take it use-after-free-merge.xml has an incorrect validate-with? If so, we should fix that, and maybe add a new regression test with an incorrect validate-with to test the handling of that (if that's possible in the PE regression test framework).

**Comment by jnpkrn** [comment]
On 30/06/16 15:29 -0700, Ken Gaillot wrote:
> I take it use-after-free-merge.xml has an incorrect validate-with?
> If so, we should fix that, and maybe add a new regression test with
> an incorrect validate-with to test the handling of that (if that's
> possible in the PE regression test framework).

No, it's correct, but that's all a minor issue I will address as well.

Important is that CIB can be constructed in such an unfortunate way
the processing part of pacemaker will just keep looping.
We should fix that ASAP (as is also signalled with `High`).

-- 
Poki

**Comment by jnpkrn** [comment]
Tweaked the commit a bit so as to account for "there was a successful validation already" case.

**Comment by kgaillot** [comment]
With your change, we can get rid of the "valid" variable altogether:

    if (validate_with(xml, lpc, to_logs) == FALSE) {
        ...
    }
    *best = lpc;
    if (transform) {
        ...
    }

However, the problem equally exists if valid==TRUE and transform==FALSE. We could add "else lpc++" after the if (transform). transform==FALSE only in cib/io.c:readCibXmlFile() if the XML validates but has no validate-with attribute, so I'm guessing that's not normally possible.

But, some branches when valid==TRUE and transform==TRUE are affected, too! I.e. "Configuration not valid for schema", "Transformation %s failed", and "Transformation %s did not produce a valid configuration" cases. So I'm thinking the safest way will be to track an orig_lpc at the beginning, and at the end, if lpc == orig_lpc, ++lpc.

Please edit the commit message to say that this fixes a bug introduced by commit 6107c63 in the 1.1.12 release. That will help me when it comes time to make the 1.1.16 changelog :)

**Comment by jnpkrn** [comment]
I am back at this, all you suggested makes sense so far, I am yet to look more at the callers, but it seems to me we should have an extra signalization for those unexpected "no validation success" cases. So far, I've extended the error output in `cli_config_update` when `version < orig_version` which helps in `crm_simulate` case, but again, need a wider look at it and see what can be done about propagation of that in general.

**Comment by jnpkrn** [comment]
In fact, "cannot validate at all" is not too interesting when the purpose is to possibly update the config, AFAICT. Nevertheless, I also did some refactoring but I guess I should add test cases to exercise different branches in `update_validation` so that we are surely good.

File: lib/common/xml.c [hunk 1]
```diff
@@ -5664,7 +5664,7 @@ update_validation(xmlNode ** xml_blob, int *best, int max, gboolean transform, g
             xmlNode *upgrade = NULL;
             int next = known_schemas[lpc].after_transform;
 
-            if (next < 0) {
+            if (next < 0 || next <= lpc) {
```
Comment (kgaillot) [inline lib/common/xml.c]: lpc is guaranteed to be >= 0 here, so we don't need next < 0

File: lib/common/xml.c [hunk 2]
```diff
@@ -5668,16 +5668,11 @@ update_validation(xmlNode ** xml_blob, int *best, int max, gboolean transform, g
                 crm_trace("Stopping at %s", known_schemas[lpc].name);
                 break;
 
-            } else if (max > 0 && lpc == max) {
+            } else if (max > 0 && (lpc == max || next > max)) {
```
Comment (kgaillot) [inline lib/common/xml.c]: Nothing to do, just an observation: 0 is a valid index, so we really should have used -1 to indicate no max

**Comment by kgaillot** [comment]
Good work. Don't worry about that next<0.

Could PE regression tests be added to verify some of the error paths?

**Pull request merged**